src/main/java/bio/terra/cli/service/utils/HttpUtils.java

```diff
@@ -108,4 +118,70 @@ public static HttpResponse sendHttpRequest(
     // return a POJO that includes both the response body and status code
     return new HttpResponse(responseBody.toString(), statusCode);
   }
+
+  /**
+   * Helper method to do retries. Uses {@link #DEFAULT_MAXIMUM_RETRIES} for maximum number of
+   * retries and {@link #DEFAULT_MILLISECONDS_SLEEP_FOR_RETRY} for the time to sleep between
+   * retries.
+   *
+   * @param makeRequest function to perform the request
+   * @param isRetryable function to test whether the exception is retryable or not
+   * @param <T> type of the Http response (i.e. return type of the makeRequest function)
+   * @return the Http response
+   * @throws Exception if makeRequest throws an exception that is not retryable, or if the maximum
+   *     number of retries was exhausted
+   */
+  public static <T> T callWithRetries(
+      HttpRequestOperator<T> makeRequest, Predicate<Exception> isRetryable) throws Exception {
+    return callWithRetries(
+        DEFAULT_MAXIMUM_RETRIES, DEFAULT_MILLISECONDS_SLEEP_FOR_RETRY, makeRequest, isRetryable);
+  }
+  /**
+   * Helper method to do retries.
+   *
+   * @param maxRetries maximum number of times to retry
+   * @param sleepDurationMS number of milliseconds to sleep between tries
+   * @param makeRequest function to perform the request
+   * @param isRetryable function to test whether the exception is retryable or not
+   * @param <T> type of the Http response (i.e. return type of the makeRequest function)
+   * @return the Http response
+   * @throws Exception if makeRequest throws an exception that is not retryable, or if the maximum
+   *     number of retries was exhausted
+   */
+  public static <T> T callWithRetries(
+      int maxRetries,
+      int sleepDurationMS,
+      HttpRequestOperator<T> makeRequest,
+      Predicate<Exception> isRetryable)
+      throws Exception {
```

Comment:
`throws Exception` isn't very helpful. Having a checked exception means we want the caller of the function to be able to handle the exception somehow. We should give them a meaningful type if we expect them to do something about it. If we don't, this function should handle what it can and then throw some runtime exception.

It looks like this is causing us to change the type to throw checked exceptions all the way up the stack. That seems like we don't actually want clients to do anything with the exception.

In this case, the Sam ApIClient calls don't actually throw checked exceptions, so we don't need to throw checked exceptions either. Our `HttpRequestOperator` doesn't need to throw a checked exception, and neither does our `callWithRetries` function. Instead, template the isRetryable with ApiException. When `makeRequest` throws an instance of that type, pass it to isRetryable. Otherwise throw the same runtime exception.
```
public static <T, E> T callWithRetries(
   ...
   Predicate<E> isRetryable) {
   ...
     try {
        logger.info("Http request attempt #{}", numTries);
        return makeRequest.makeRequest();
      } catch (RuntimeException ex) {
        // if the exception is not retryable, then quit polling
        if (!(ex instanceof E) || !isRetryable.test((E) ex)) {
          throw ex;
        }
        logger.info("Caught retryable exception: {}", ex);
      }
 ```

Reply:
I'm confused about "the Sam ApIClient calls don't actually throw checked exceptions". All the Sam ApiClient calls throw ApiException, which is a checked exception.

Reply:
Sorry, that's right. Switching to throwing the ApiException lgtm.